### ISSUE TITLE
[Accessibility] Device Dependent handlers are used. Tooltips are not accessible with keyboard.

### ISSUE THREAD

**Issue by puneet-ahuja** [vega/vega#3878]
Tooltips in the charts are not accessible with the help of a keyboard.

Steps to replicate:

Open the [Codesandbox Link](https://codesandbox.io/p/sandbox/tooltip-not-accessible-with-keyboard-yh3ldj?file=%2Findex.html%3A63%2C8).

The above link opens an interactive multi-line chart.
We can access the tooltips on the lines using mouse hover but not with keyboards.

**Comment by domoritz** [comment]
Yes, that's an issue. The tricky thing is that we don't always have navigable elements since Vega can render to Canvas. So this may not be something that can be solved in general but we should look into having keyboard navigation with the SVG renderer at least. 

For Canvas, we could suggest people use something like https://github.com/cmudig/data-navigator (cc @frankelavsky).

**Comment by majornista** [comment]
@domoritz I agree with you that in the same sense that aria props, like role and description, are only relevant when rendering with SVG, keyboard navigation should work within the SVG renderer. Deciding on how to implement keyboard navigation can be tricky and requires flexibility within the renderer. For example, with many datapoints in a multiline chart, you probably don't want every datapoint that renders a tooltip on hover to be included in the tab order, but instead provide some mechanism for "drilling into" and escaping back out of different levels of granularity. For example, in a multiline stock chart, you might want to focus the line for one of the stocks or indexes displayed and get a description of the general trend, before drilling in to get detail on every datapoint or trade. @frankelavsky does something along these lines in the [Data Navigator demo ](https://dig.cmu.edu/data-navigator/) of a stacked bar chart.

**Comment by majornista** [comment]
@frankelavsky and @domoritz Thanks for joining the conversation. I agree that if it's not easy to define these types of navigation patterns, developers won't do it. My first inclination, as someone who spends a great deal of time doing accessibility remediation, would be to dive in and try to fill the feature gaps that prevent a developer from being able to author an accessible experience, like supporting `tabindex` for marks, as was your original inclination in #2548, so that a developer could potentially define focus and keyboard events for marks. But, even if the grammar were to support these features, the amount of effort for someone using Vega or Vega-lite to craft an accessible experience would be still be high, and may even require defining a bespoke navigational experience for each data viz. 

We should strive for something that works well for most use cases out of the box, like the keyboard accessibility for Tooltips requested in this issue, yet can be customized to support more advanced features, like making the legend behave as radio buttons, as in [this Interactive Legend example](https://vega.github.io/editor/#/examples/vega/interactive-legend).

**Comment by majornista** [comment]
@domoritz and @frankelavsky I have been doing some experimentation to add support for `tabindex` and `focusin`/`focusout` events when rendering to SVG. While I can support something as simple as updating the style for the focused element or displaying a tooltip (I started small, experimenting with bar-chart.vg.json and airport-connections.vg.json examples), anything more complex, like keyboard navigation through elements using the arrow keys, or drilling down into or up out of a group of elements, like moving from a line to the points within it, will require some way to manage focus within the rendered SVG. I think this may requires a bit more "programming" than the current declarative syntax supports. For example, while I can detect that a user has key pressed ArrowLeft or ArrowRight on an element, using the event.target. I can't figure out and set focus to the appropriate adjacent element from within a Vega spec. 

Any suggestions on how this might be implemented within the Vega spec would be greatly appreciated.

**Comment by domoritz** [comment]
I'd say make the generated svg accessible first without worrying about tooltips. The default tooltip handler is just setting the title property iirc. Vega should provide the basic building blocks to make a chart accessible but not do smart things by default. Smartness can come in in Vega-Lite which generates Vega specs. 

To make the tooltips more accessible, we need to modify Vega tooltip which is an implementation of the handler that renders tooltips as html. I'm saying to do this later since people can implement their own handlers and tooltips are not as important if we have good labels on the chart elements in Vega.

**Comment by majornista** [comment]
> I'd say make the generated svg accessible first without worrying about tooltips. The default tooltip handler is just setting the title property iirc. Vega should provide the basic building blocks to make a chart accessible but not do smart things by default. Smartness can come in in Vega-Lite which generates Vega specs.
> 
> To make the tooltips more accessible, we need to modify Vega tooltip which is an implementation of the handler that renders tooltips as html. I'm saying to do this later since people can implement their own handlers and tooltips are not as important if we have good labels on the chart elements in Vega.

While many of the examples don't do a great job of providing a meaningful `description` or `roleDescription` at the moment, it is possible to add them, and theoretically the marks can be made accessible to screen readers. For example, in the airport-connections.vg.json example, I can add the following to each airport symbol to render something meaningful to a screen reader user:

```json
          "ariaRoleDescription": {"value": "airport mark symbol"},
          "description": {"signal": "datum.name + ' (' + datum.iata + '), ' + datum.city + ', ' + datum.state + ' (' + datum.traffic.flights + ' flights in 2008)'"},
```

However, a when a mouse user hovers an airport, they are able to see a rendering of all the connecting routes made to that airport, and this can be defined as part of the Vega visualization. A keyboard-only user cannot access this functionality, which creates is a Level A conformance failure of [WCAG 2.2 Success Criterion 2.1.1 Keyboard](https://www.w3.org/TR/WCAG22/#keyboard), and Vega currently offers no suitable way to work around it. With the changes, I've been working on, which adds support for `tabindex` and `focusin`/`focusout` event handling on marks, I can make each airport node focusable using `tabindex="0"`, and hide/show the connecting routes on focus, but this causes every airport in the visualization to be included in the keyboard navigation order for the web page, which with hundreds of nodes, creates a tedious keyboard navigation experience. 

There are a couple ways to make this keyboard experience more humane, but without support for focus management within the Vega spec, there does not seem to be a way to implement them:

1. Use roving `tabindex`, so that only one airport is included in the tab order at a time, and the user navigates between airports using the arrow keys.
2. Provide a mechanism to drill down to a more granular navigation level. The user first focuses a container element with a description, then by pressing enter/space or clicking, they drill down into a navigation mode to access details for each airport, which can then be navigated with tab or shift+tab keys. Pressing Esc, escapes the navigation mode, restoring focus to the container element, at which point the user can tab navigate away from the visualization without having to navigate through hundreds of nodes.

Both of these options seem to require an ability to manage focus between marks within a Vega spec.

**Comment by majornista** [comment]
I made some experimental progress adding `tabindex`, support for `focusin`/`focusout` events, a `focus(element)` event function that allows one to focus an SVG element within the scenegraph:

https://github.com/majornista/vega/commit/9517f5c77097ef9f464b9773b88e7d1bff178244

These changes allow bars in the bar-chart.vg.json example and airports within the airport-connections.vg.json example to be keyboard navigable using roving tabindex.
